src/java/voldemort/store/readonly/ReadOnlyFileDiffer.java

```diff
@@ -0,0 +1,383 @@
+     * @return result of difference. Including number of added, deleted, updated and the same.
+     */
+    protected long[] comparePartition(ByteArrayComparator comparator, ReadOnlyRecordIterator oldIter,
+        ReadOnlyRecordIterator newIter) {
+        long[] results = new long[4];
+        byte[] oldIndex;
+        byte[] newIndex;
+
+        oldIndex = oldIter.next();
+        newIndex = newIter.next();
+
+        while (oldIndex != null && newIndex != null) {
+            int result = comparator.compare(oldIndex, newIndex);
+            if (result == 0) {
+                //Same hash of keys. Compare keys and values more.
+                compareRecordsWithSameHash(comparator, oldIter.readRecords(), newIter.readRecords(), results);
+                oldIndex = oldIter.next();
+                newIndex = newIter.next();
+            } else if (result > 0) {
+                //old hash> new hash.
+                results[ADD]++;
+                newIndex = newIter.next();
+            } else {
+                //old hash < new hash.
+                results[DELETE]++;
+                oldIndex = oldIter.next();
+            }
+        }
+
+        while (oldIndex != null) {
+            results[DELETE]++;
+            oldIndex = oldIter.next();
+        }
+
+        while (newIndex != null) {
+            results[ADD]++;
+            newIndex = newIter.next();
+        }
+        return results;
+    }
+
+    /**
+     * Compare the keys and values one by one between two different versions.
+     */
+    private void compareRecordsWithSameHash(ByteArrayComparator comparator, Pair<byte[], byte[]>[] oldRecords,
+        Pair<byte[], byte[]>[] newRecords, long[] results) {
+        int numOldRecords = oldRecords.length;
+        int numNewRecords = newRecords.length;
+        // These records all have the same hash value, thus maybe there is not fixed order here. \
+        // Just compare them one by one as the number of records should be very small.
+        for (Pair<byte[], byte[]> oldPRecord : oldRecords) {
+            for (Pair<byte[], byte[]> newRecord : newRecords) {
+                if (comparator.compare(oldPRecord.getFirst(), newRecord.getFirst()) == 0) {
+                    numOldRecords--;
+                    numNewRecords--;
+                    if (comparator.compare(oldPRecord.getSecond(), newRecord.getSecond()) == 0) {
+                        results[SAME]++;
+                    } else {
+                        results[UPDATE]++;
+                    }
+                }
+            }
+        }
+        results[ADD] += numNewRecords;
+        results[DELETE] += numOldRecords;
+    }
+
+    static class ByteArrayComparator implements Comparator<byte[]> {
+
+        @Override
+        public int compare(byte[] o1, byte[] o2) {
+            int i = 0;
+            int j = 0;
+            for (; i < o1.length && j < o2.length; i++, j++) {
+                if (o1[i] < o2[i]) {
```

Comment:
AFAIK, java's byte is signed type. Ranging from -128 to 127. I don't know whether it would cause problem in comparison. But I would prefer the byte array comparator to use one of the proven methods instead of we writing one from scratch. 

http://stackoverflow.com/questions/5108091/java-comparator-for-byte-array-lexicographic